File: ai_visual_tester/tools.py
from __future__ import annotations

from typing import Any


def get_tool_declarations() -> list[dict[str, Any]]:
    """Return Gemini tool/function declarations for common UI actions.

    The schema follows Gemini's function declaration format.
    """
    return [
        {
            "function_declarations": [
                {
                    "name": "move_cursor",
                    "description": "Move the mouse cursor to absolute screen coordinates.",
                    "parameters": {
                        "type": "object",
                        "properties": {
                            "x": {"type": "integer", "minimum": 0},
                            "y": {"type": "integer", "minimum": 0},
                        },
                        "required": ["x", "y"],
                    },
                },
                {
                    "name": "click",
                    "description": "Click at absolute screen coordinates.",
                    "parameters": {
                        "type": "object",
                        "properties": {
                            "x": {"type": "integer", "minimum": 0},
                            "y": {"type": "integer", "minimum": 0},
                            "button": {
                                "type": "string",
                                "enum": ["left", "right", "middle"],
                                "default": "left",
                            },
                            "clicks": {"type": "integer", "minimum": 1, "default": 1},
                        },
                        "required": ["x", "y"],
                    },
                },
                {
                    "name": "drag_and_drop",
                    "description": "Drag from (start_x, start_y) and drop at (end_x, end_y).",
                    "parameters": {
                        "type": "object",
                        "properties": {
                            "start_x": {"type": "integer", "minimum": 0},
                            "start_y": {"type": "integer", "minimum": 0},
                            "end_x": {"type": "integer", "minimum": 0},
                            "end_y": {"type": "integer", "minimum": 0},
                        },
                        "required": ["start_x", "start_y", "end_x", "end_y"],
                    },
                },
                {
                    "name": "type_text",
                    "description": "Type the given text at the current focus.",
                    "parameters": {
                        "type": "object",
                        "properties": {
                            "text": {"type": "string"},
                            "enter": {"type": "boolean", "default": False},
                        },
                        "required": ["text"],
                    },
                },
                {
                    "name": "key_press",
                    "description": "Press a single key or a chord (e.g., CTRL+C).",
                    "parameters": {
                        "type": "object",
                        "properties": {
                            "keys": {
                                "type": "array",
                                "items": {"type": "string"},
                                "description": "Ordered list of keys to press as a chord.",
                            }
                        },
                        "required": ["keys"],
                    },
                },
                {
                    "name": "scroll",
                    "description": "Scroll vertically by a delta in pixels (positive is down).",
                    "parameters": {
                        "type": "object",
                        "properties": {
                            "delta_y": {"type": "integer"},
                        },
                        "required": ["delta_y"],
                    },
                },
                {
                    "name": "wait",
                    "description": "Wait for a number of seconds to allow UI updates.",
                    "parameters": {
                        "type": "object",
                        "properties": {
                            "seconds": {"type": "number", "minimum": 0},
                        },
                        "required": ["seconds"],
                    },
                },
                {
                    "name": "assert_text_present",
                    "description": "Assert that the given text should be visible on screen.",
                    "parameters": {
                        "type": "object",
                        "properties": {
                            "text": {"type": "string"},
                        },
                        "required": ["text"],
                    },
                },
            ]
        }
    ]

File: src/tools.js
"use strict";

// Tool declarations for Gemini function calling and local executor implementations

const toolDeclarations = [
  {
    name: "move_cursor",
    description: "Move the mouse cursor to the given screen coordinates (pixels).",
    parameters: {
      type: "object",
      properties: {
        x: { type: "number", description: "X coordinate in pixels from left" },
        y: { type: "number", description: "Y coordinate in pixels from top" }
      },
      required: ["x", "y"]
    }
  },
  {
    name: "click",
    description: "Click at the given coordinates. If coordinates are omitted, click at current cursor position.",
    parameters: {
      type: "object",
      properties: {
        x: { type: "number", description: "X coordinate in pixels from left" },
        y: { type: "number", description: "Y coordinate in pixels from top" },
        button: {
          type: "string",
          description: "Mouse button",
          enum: ["left", "right", "middle"],
          default: "left"
        }
      }
    }
  },
  {
    name: "type_text",
    description: "Type the provided text at the current focus location.",
    parameters: {
      type: "object",
      properties: {
        text: { type: "string", description: "The text to type" }
      },
      required: ["text"]
    }
  },
  {
    name: "scroll",
    description: "Scroll by the given deltas. Positive deltaY scrolls down.",
    parameters: {
      type: "object",
      properties: {
        deltaX: { type: "number", description: "Horizontal scroll delta in pixels", default: 0 },
        deltaY: { type: "number", description: "Vertical scroll delta in pixels", default: 0 }
      }
    }
  },
  {
    name: "wait_for",
    description: "Wait for a UI condition or a fixed delay before continuing.",
    parameters: {
      type: "object",
      properties: {
        ms: { type: "number", description: "Milliseconds to wait" },
        note: { type: "string", description: "What are we waiting for?" }
      }
    }
  },
  {
    name: "end_test",
    description: "Finish the test and provide a concise summary and pass/fail status.",
    parameters: {
      type: "object",
      properties: {
        summary: { type: "string", description: "1-3 sentence summary of what happened and the result" },
        passed: { type: "boolean", description: "Did the test objective succeed?" },
        issues: { type: "array", items: { type: "string" }, description: "List of issues or blockers encountered", default: [] }
      },
      required: ["summary", "passed"]
    }
  }
];

// Simple local executor: logs and returns acknowledgements. Replace with real drivers (e.g., Playwright) as needed.
const actionLog = [];

function recordAction(action) {
  const timestamp = new Date().toISOString();
  const entry = { timestamp, ...action };
  actionLog.push(entry);
  // Keep the log from exploding in long sessions
  if (actionLog.length > 1000) actionLog.shift();
  return entry;
}

const localToolExecutor = {
  async move_cursor(args) {
    const { x, y } = args || {};
    if (typeof x !== "number" || typeof y !== "number") {
      return { ok: false, error: "x and y are required numbers" };
    }
    const entry = recordAction({ tool: "move_cursor", x, y });
    return { ok: true, entry };
  },
  async click(args) {
    const { x, y, button = "left" } = args || {};
    const entry = recordAction({ tool: "click", x, y, button });
    return { ok: true, entry };
  },
  async type_text(args) {
    const { text } = args || {};
    if (typeof text !== "string") {
      return { ok: false, error: "text is required" };
    }
    const entry = recordAction({ tool: "type_text", text });
    return { ok: true, entry };
  },
  async scroll(args) {
    const { deltaX = 0, deltaY = 0 } = args || {};
    const entry = recordAction({ tool: "scroll", deltaX, deltaY });
    return { ok: true, entry };
  },
  async wait_for(args) {
    const { ms = 500, note } = args || {};
    const entry = recordAction({ tool: "wait_for", ms, note });
    if (ms > 0) {
      await new Promise((resolve) => setTimeout(resolve, Math.min(ms, 2000)));
    }
    return { ok: true, entry };
  },
  async end_test(args) {
    const { summary, passed, issues = [] } = args || {};
    const entry = recordAction({ tool: "end_test", summary, passed, issues });
    return { ok: true, entry };
  }
};

module.exports = {
  toolDeclarations,
  localToolExecutor,
  actionLog
};

